File: 프로그래머스/lv2.다리를 지나는 트럭.cpp
// Level2: 다리를 지나는 트럭

#include <bits/stdc++.h>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    queue<int> q;
    for(int i = 0; i < bridge_length; i++) q.push(0);
    int w = 0, idx = 0;

    while(!q.empty()){
        answer ++;
        w -= q.front();
        q.pop();
        if(w + truck_weights[idx] <= weight){
            q.push(truck_weights[idx]);
            w += truck_weights[idx];
            idx ++;
            if(idx == truck_weights.size())
                return  answer + bridge_length;
        }
        else if (w + truck_weights[idx] > weight) {
            q.push(0);
        }
    }
}

File: 프로그래머스/lv2.다리를 지나는 트럭.js
// Level: 다리를 지나는 트럭
function solution(bridge_length, weight, truck_weights) {
    let answer = 0;
    let queue = [];
    let cur_weight = 0;
    let cur_idx = 0;
    for(let i = 0; i < bridge_length; i++){
        queue.push(0);
    }
    while(cur_idx < truck_weights.length){
        answer += 1;
        cur_weight -= queue[0];
        queue.shift();
        if(cur_weight + truck_weights[cur_idx] <= weight){
            queue.push(truck_weights[cur_idx]);
            cur_weight += truck_weights[cur_idx];
            cur_idx += 1;
        }else {
            queue.push(0);
        }
    }
    answer += bridge_length;
    return answer;
}